src/generate_conf.rs

```diff
@@ -0,0 +1,228 @@
+use std::ffi::OsStr;
+use std::fs;
+use std::path::Path;
+
+use anyhow::{anyhow, Context};
+use log::{info, warn};
+use nmstate::{InterfaceType, NetworkState};
+use serde::{Deserialize, Serialize};
+
+const HOST_MAPPING_FILE: &str = "host_config.yaml";
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub struct HostInterfaces {
+    hostname: String,
+    interfaces: Vec<Interface>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub struct Interface {
+    logical_name: String,
+    mac_address: String,
+}
+
+/// NetworkConfig contains the generated configurations in the
+/// following format: Vec<(config_file_name, config_content>)
+type NetworkConfig = Vec<(String, String)>;
+
+/// Generate network configurations from all YAML files in the `config_dir`
+/// and store the result *.nmconnection files and host mapping under `output_dir`.
+pub(crate) fn generate(config_dir: &str, output_dir: &str) -> Result<(), anyhow::Error> {
+    for entry in fs::read_dir(config_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if entry.metadata()?.is_dir() {
+            warn!("Ignoring unexpected dir: {path:?}");
+            continue;
+        }
+
+        info!("Generating config from {path:?}...");
+
+        let hostname = path
+            .file_stem()
+            .and_then(OsStr::to_str)
+            .ok_or_else(|| anyhow!("Invalid file path"))?
+            .to_string();
+
+        let data = fs::read_to_string(&path).with_context(|| "Reading network config")?;
+
+        let (interfaces, config) = generate_config(data)?;
+
+        store_network_config(output_dir, hostname, interfaces, config)
+            .with_context(|| "Storing config")?;
+    }
+
+    Ok(())
+}
+
+fn generate_config(data: String) -> Result<(Vec<Interface>, NetworkConfig), anyhow::Error> {
+    let network_state = NetworkState::new_from_yaml(&data)?;
+
+    let interfaces = extract_interfaces(&network_state);
+    let config = network_state
+        .gen_conf()?
+        .get("NetworkManager")
+        .ok_or_else(|| anyhow!("Invalid NM configuration"))?
+        .to_owned();
+
+    Ok((interfaces, config))
+}
+
+fn extract_interfaces(network_state: &NetworkState) -> Vec<Interface> {
+    network_state
+        .interfaces
+        .iter()
+        .filter(|i| i.iface_type() != InterfaceType::Loopback)
+        .filter(|i| i.base_iface().mac_address.is_some())
+        .map(|i| Interface {
+            logical_name: i.name().to_string(),
+            mac_address: i.base_iface().mac_address.clone().unwrap(),
+        })
+        .collect()
+}
+
+fn store_network_config(
+    output_dir: &str,
+    hostname: String,
+    interfaces: Vec<Interface>,
+    config: NetworkConfig,
+) -> Result<(), anyhow::Error> {
+    let path = Path::new(output_dir);
+
+    fs::create_dir_all(path.join(&hostname)).with_context(|| "Creating output dir")?;
```

Comment:
you should use the `context` method for this as this is a static value

Reply:
thanks!